axlearn/common/checkpointer_orbax_emergency.py

```diff
@@ -0,0 +1,796 @@
+    jax.distributed.shutdown()
+
+
+def get_consistent_proc_info(
+    *,
+    local_address: Optional[str] = None,
+    barrier_timeout_seconds: int = 300,
+    trainer_dir: str,
+    local_ckpt_dir: str,
+    **setup_kwargs,
+) -> _ProcessInfo:
+    """Gets the invariant process id of the current process and global coordinator's address.
+
+    This function guarantees process id <-> node mapping stays the same for healthy nodes after a
+    failover. This is required to preserve shard order for in-memory checkpoint recovery. For GPU
+    training, all healthy nodes will have their process id unchanged. For TPU, all nodes in the
+    healthy slices will have their process id unchanged. See docstring of
+    `_init_consistent_proc_ids` for implementation details.
+
+    Args:
+        local_address: A IP:Port that can be used as the coordinator if this rank is elected.
+            This Port must be free in the coordinator pod and IP:Port must be reachable from all
+            other processes.
+        barrier_timeout_seconds: Timeout in seconds for the barrier and key_value_set operations.
+        trainer_dir: Path to the trainer dir.
+        local_ckpt_dir: Path to the local checkpoint dir.
+        **setup_kwargs: Args to `utils_spmd.setup()`.
+
+    Returns:
+        A _ProcessInfo whose `inv_proc_id` should be used as the process id and `address` should be
+        used as the global coordinator address.
+    """
+    platform = os.environ.get("JAX_PLATFORMS", "")
+    try:
+        start_t = time.perf_counter()
+        # Patch platform so the process doesn't waste time initializing accelerators.
+        os.environ["JAX_PLATFORMS"] = "cpu"
+        proc = mp.get_context("spawn").Process(
+            target=_init_consistent_proc_ids,
+            kwargs=dict(
+                local_address=local_address,
+                barrier_timeout_seconds=barrier_timeout_seconds,
+                trainer_dir=trainer_dir,
+                local_ckpt_dir=local_ckpt_dir,
+                **setup_kwargs,
+            ),
+        )
+        proc.start()
+        proc.join()
+        if proc.exitcode != 0:
+            raise RuntimeError(
+                "Expects id assignment process to finish normally. "
+                f"Got exit code {proc.exitcode}. Please check the log above for errors."
+            )
+
+        info = _get_previous_process_info(local_ckpt_dir, trainer_dir=trainer_dir)
+        if info.inv_proc_id == -1:
+            raise RuntimeError("Expects inv process id != -1, but got -1.")
+        logging.info(
+            "Successfully finished process ID assignment in %fs", time.perf_counter() - start_t
+        )
+        return info
+    finally:
+        # Restore previous platform settings.
+        if platform != "":
+            os.environ["JAX_PLATFORMS"] = platform
+        else:
+            del os.environ["JAX_PLATFORMS"]
+
+
+class OrbaxEmergencyCheckpointer(BaseCheckpointer):
```

Comment:
How does this overlap with https://github.com/apple/axlearn/blob/140a18f33d10178dc695469e36d39d7879bd1ed4/axlearn/common/checkpointer_orbax.py#L169?

Reply:
The orbax regular checkpointer saves 1 gcs checkpoint for n slices per save. This checkpointer saves n-1 checkpoints to a local path (usually a ramdisk), and also 1 checkpoint to gcs.

Reply:
Unfortunately it's not possible to share code between the two implementations.

Reply:
Thanks for the clarification.
- How should users use them? Should they use both of them or one of them? How should they pick?
- Or can we replace OrbaxCheckpointer with this class?

Reply:
Added a comment to clarify this:
```
    This checkpointer is designed to improve the goodput of large multi-slice training jobs that
    use data-parallelism across slices. At least two data-parallel slices are required. For other
    use cases where this is not applicable or ultimate goodput is not required, please use
    `OrbaxCheckpointer`.

```

Reply:
OK. So we will need to maintain two separate classes going forward? That doesn't seem ideal.

Could you elaborate on why we need two completely separate implementations?

Reply:
I think the fundamental reason is that orbax emergency checkpointer doesn't support non-tensor states, so we have to have a different checkpointer wrapper around it.

Reply:
I see. What prevents us from using the emergency checkpointer for all jobs?

Reply:
It requires two data-parallel slices

Reply:
Not just data-parallelism, but data-parallel slices, i.e. it has to be multi-slice training.

Reply:
IIUC, only the local checkpoints require multiple slices, since we will need to restore from another slice upon a slice restart. Could we disable local checkpoints when num_slices=1? This way we always use the emergency checkpointer consistently.

Reply:
Could be an idea.

Reply:
Still needs support from orbax though.

Reply:
I see. Right now it assumes local checkpoints are always used: https://github.com/google/orbax/blob/6e80ecc27581a413b1a481d4740e61df7316a4f4/checkpoint/orbax/checkpoint/experimental/emergency/checkpoint_manager.py#L695-L709.

Could you raise this request to the Orbax team and link to the issue?

Reply:
Done.